packages/opentelemetry-tracer/src/instrumentation/PluginLoader.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Tracer, Logger, Plugin } from '@opentelemetry/types';
+import * as utils from './utils';
+import * as hook from 'require-in-the-middle';
+
+// States for the Plugin Loader
+export enum HookState {
+  UNINITIALIZED,
+  ENABLED,
+  DISABLED,
+}
+
+interface PluginNames {
+  [pluginName: string]: string;
+}
+
+// tslint:disable-next-line:no-any
+export type PluginType = Array<Plugin<any>>;
+
+/**
+ * The PluginLoader class can load instrumentation plugins that
+ * use a patch mechanism to enable automatic tracing for
+ * specific target modules.
+ */
+export class PluginLoader {
+  /** A list of loaded plugins. */
+  private _plugins: PluginType = [];
+  /**
+   * A field that tracks whether the require-in-the-middle hook has been loaded
+   * for the first time, as well as whether the hook body is activated or not.
+   */
+  private _hookState = HookState.UNINITIALIZED;
+
+  /** Constructs a new PluginLoader instance. */
+  constructor(readonly tracer: Tracer, readonly logger: Logger) {}
+
+  get state(): HookState {
```

Comment:
What is the use case to expose this?

Reply:
`state()` and `plugins()` methods are used for testing only.

Reply:
I'm not a fan of exposing public properties if they are going to be used only for testing. Is there a way that these would not be needed, or to not actually expose them publicly?

Reply:
What's your suggestion?

Reply:
From a code perspective, one solution could be to move them to a separate file (like plugin-loader-consts.js`) and then you can require them in both the test and in the class without making it public.

Reply:
From a JavaScript perspective the private properties could simply be used in the test. From a TypeScript perspective it could be stored on a `WeakMap` that lives in another file which could be imported in the test but not otherwise exposed.

Reply:
@hekike and @rochdev thanks for the feedback and suggestions. Done in https://github.com/open-telemetry/opentelemetry-js/pull/126/commits/16cd6c9629eda0c7ae7a4bee67067f3e3ca228a9